README.md
# Drop Link

A Flutter Web plugin to handle drag-and-drop (files) into Flutter. If you're interested in drag-and-drop inside a Flutter app, check out other packages like dnd.

* [Getting Started](#getting-started)
* [Installation](#installation)
* [Usage](#usage)
* [Example App](https://github.com/KalybosPro/drop_link/tree/main/example)

## Getting Started

This flutter plugin help to drop link in flutter web app.

## Installation

This plugin is available on Pub: [https://pub.dev/packages/drop_link](https://pub.dev/packages/drop_link)

Add this to `dependencies` in your app's `pubspec.yaml`

```yaml
drop_link: ^1.0.1
```

## Usage

Sample code to integrate can be found in [example/lib/main.dart](example/lib/main.dart).

#### Import package

```dart
import 'package:drop_link/drop_link.dart';
```

#### Example

```dart
import 'package:dotted_border/dotted_border.dart';
import 'package:drop_link/drop_link.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Drop link zone',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
   String? link;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Drop link zone"),
      ),
      body: Center(
        child: Container(
          height: 100,
          width: 500,
          alignment: Alignment.center,
          child: DropLinkZone(
            onDrop: (data) {
              debugPrint(data);
              setState(() {
                link = data;
              });
            },
            child: DottedBorder(
              color: Colors.black.withOpacity(0.4),
              strokeWidth: 2,
              dashPattern: const [4, 2],
              radius: const Radius.circular(3),
              child: Container(
                alignment: Alignment.center,
                decoration: const BoxDecoration(
                  color: Colors.white,
                ),
                child: link!=null? Center(child: Text("$link")): Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: const [
                    Spacer(),
                    ClipRRect(
                      child:  Icon(Icons.link),
                    ),
                    Spacer(),
                    Text("Drop your link here"),
                    Spacer(),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```
#### Screenshots

<img src="https://github.com/KalybosPro/drop_link/blob/main/Capture%20web_8-9-2023_13262_localhost.jpeg" height="500px"/>
<img src="https://github.com/KalybosPro/drop_link/blob/main/Capture%20web_8-9-2023_133143_localhost.jpeg" height="500px"/>

CHANGELOG.md
## 1.0.1

* This version can work for Flutter web app only.


test/drop_link_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:drop_link/drop_link.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/drop_zone.dart
// ignore: avoid_web_libraries_in_flutter
import 'dart:html' as html;

import 'package:flutter/material.dart';

import 'dispatcher.dart';

typedef VoidCallback<T> = void Function(T);

class DropLinkZone extends StatefulWidget {
  final Widget child;
  final void Function()? onDragEnter;
  final void Function()? onDragExit;

  final VoidCallback<String>? onDrop;

  const DropLinkZone({
    required this.child,
    Key? key,
    this.onDrop,
    this.onDragEnter,
    this.onDragExit,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => DropLinkZoneState();
}

class DropLinkZoneState extends State<DropLinkZone> {
  late int id;
  bool _dragInBounds = false;

  Rect? _getGlobalPaintBounds() {
    final renderObject = context.findRenderObject();
    final translation = renderObject?.getTransformTo(null).getTranslation();

    if (translation != null && renderObject?.paintBounds != null) {
      return renderObject?.paintBounds
          .shift(Offset(translation.x, translation.y));
    }

    return null;
  }

  @override
  void dispose() {
    Dispatcher.shared.removeZone(id);

    super.dispose();
  }

  bool _isCursorWithinBounds(html.MouseEvent value) {
    final g = _getGlobalPaintBounds();

    return g != null && //
        value.layer.x > g.left &&
        value.layer.x < g.right &&
        value.layer.y > g.top &&
        value.layer.y < g.bottom;
  }

  @override
  void initState() {
    super.initState();

    id = Dispatcher.shared.addZone(onDragOver: (e) {
      final tmp = _isCursorWithinBounds(e);
      if (!_dragInBounds && tmp) {
        widget.onDragEnter?.call();
      } else if (_dragInBounds && !tmp) {
        widget.onDragExit?.call();
      }

      _dragInBounds = tmp;
    }, onDrop: (e) async {
      if (_isCursorWithinBounds(e)) {
        final data = e.dataTransfer.getData("text");
        widget.onDrop?.call(data);
      }
    });
  }

  @override
  Widget build(BuildContext c) => widget.child;
}


example/lib/main.dart
import 'package:dotted_border/dotted_border.dart';
import 'package:drop_link/drop_link.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Drop link zone',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
   String? link;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Drop link zone"),
      ),
      body: Center(
        child: Container(
          height: 100,
          width: 500,
          alignment: Alignment.center,
          child: DropLinkZone(
            onDrop: (data) {
              debugPrint(data);
              setState(() {
                link = data;
              });
            },
            child: DottedBorder(
              color: Colors.black.withOpacity(0.4),
              strokeWidth: 2,
              dashPattern: const [4, 2],
              radius: const Radius.circular(3),
              child: Container(
                alignment: Alignment.center,
                decoration: const BoxDecoration(
                  color: Colors.white,
                ),
                child: link!=null? Center(child: Text("$link")): Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: const [
                    Spacer(),
                    ClipRRect(
                      child:  Icon(Icons.link),
                    ),
                    Spacer(),
                    Text("Drop your link here"),
                    Spacer(),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
